src/main/java/com/gregtechceu/gtceu/api/recipe/GTRecipe.java

```diff
@@ -159,72 +156,36 @@ public List<Content> getTickOutputContents(RecipeCapability<?> capability) {
     }
 
     public ActionResult matchRecipe(IRecipeCapabilityHolder holder) {
+        return matchRecipe(holder, false);
+    }
+
+    public ActionResult matchTickRecipe(IRecipeCapabilityHolder holder) {
+        return hasTick() ? matchRecipe(holder, true) : ActionResult.SUCCESS;
+    }
+
+    private ActionResult matchRecipe(IRecipeCapabilityHolder holder, boolean tick) {
         if (!holder.hasProxies()) return ActionResult.FAIL_NO_REASON;
-        var result = matchRecipe(IO.IN, holder, inputs, false);
+
+        var result = matchRecipeContents(IO.IN, holder, tick ? tickInputs : inputs);
         if (!result.isSuccess()) return result;
-        result = matchRecipe(IO.OUT, holder, outputs, false);
+
+        result = matchRecipeContents(IO.OUT, holder, tick ? tickOutputs : outputs);
         if (!result.isSuccess()) return result;
-        return ActionResult.SUCCESS;
-    }
 
-    public ActionResult matchTickRecipe(IRecipeCapabilityHolder holder) {
-        if (hasTick()) {
-            if (!holder.hasProxies()) return ActionResult.FAIL_NO_REASON;
-            var result = matchRecipe(IO.IN, holder, tickInputs, false);
-            if (!result.isSuccess()) return result;
-            result = matchRecipe(IO.OUT, holder, tickOutputs, false);
-            if (!result.isSuccess()) return result;
-        }
         return ActionResult.SUCCESS;
     }
 
-    public ActionResult matchRecipe(IO io, IRecipeCapabilityHolder holder, Map<RecipeCapability<?>, List<Content>> contents, boolean calculateExpectingRate) {
-        Table<IO, RecipeCapability<?>, List<IRecipeHandler<?>>> capabilityProxies = holder.getCapabilitiesProxy();
+    public ActionResult matchRecipeContents(IO io, IRecipeCapabilityHolder holder, Map<RecipeCapability<?>, List<Content>> contents) {
         for (Map.Entry<RecipeCapability<?>, List<Content>> entry : contents.entrySet()) {
-            Set<IRecipeHandler<?>> used = new HashSet<>();
-            List content = new ArrayList<>();
-            Map<String, List> contentSlot = new HashMap<>();
-            for (Content cont : entry.getValue()) {
-                if (cont.slotName == null) {
-                    content.add(cont.content);
-                } else {
-                    contentSlot.computeIfAbsent(cont.slotName, s -> new ArrayList<>()).add(cont.content);
-                }
-            }
-            RecipeCapability<?> capability = entry.getKey();
-            if (!capability.doMatchInRecipe()) {
+            var result = new RecipeRunner(this, io, holder, true).handle(entry);
```

Comment:
if you create a new `RecipeRunner` and only ever use it once, why do you even need to create an instance of it? why not just have these methods be static?

Reply:
Because that makes the method signatures **a lot** smaller and easier to understand.

The instance state is currently just shared data between the individual methods, but I'm planning to do some more refactoring there, to reuse the same runner across all entries.